src/gsi/hybrid_ensemble_isotropic.F90

```diff
@@ -3938,53 +4056,162 @@ subroutine ckgcov_a_en_new_factorization_ad(ig,z,a_en)
      call stop2(999)
   endif
 
+! MGBF-based localization (now available only in regional=.true.)
+  if(l_mgbf_loc) then
+
 ! Apply vertical smoother on each ensemble member
-  iadvance=1 ; iback=2
-!$omp parallel do schedule(static,1) private(k)
-  do k=1,n_ens
+     allocate(work_mgbf(obj_mgbf(ig)%km_a_all,obj_mgbf(ig)%nm,obj_mgbf(ig)%mm))
+     work_mgbf=zero
+     iadvance=1 ; iback=2
+!$omp parallel do schedule(static,1) private(k,ii,is,ie)
+     do k=1,n_ens
+        ii=(k-1)*grd_loc%nsig
+        is=ii+1
+        ie=ii+grd_loc%nsig
+        if(.not.obj_mgbf(ig)%l_vertical_filter) call new_factorization_rf_z(a_en(k)%r3(ipnt)%q,iadvance,iback,ig)
+        call map_work_mgbf(a_en(k)%r3(ipnt)%q,work_mgbf(is:ie,:,:),iadvance,ig)
+     enddo
 
-     call new_factorization_rf_z(a_en(k)%r3(ipnt)%q,iadvance,iback,ig)
- 
-  enddo
+! Mapping from analysis grid to filter grid
+     if(obj_mgbf(ig)%km_a_all==obj_mgbf(ig)%km_all &
+          &.and. obj_mgbf(ig)%nm==obj_mgbf(ig)%im &
+          &.and. obj_mgbf(ig)%mm==obj_mgbf(ig)%jm) then
+        obj_mgbf(ig)%VALL=zero
+        obj_mgbf(ig)%VALL(1:obj_mgbf(ig)%km_all,1:obj_mgbf(ig)%im,1:obj_mgbf(ig)%jm)=work_mgbf
+     elseif(obj_mgbf(ig)%l_new_map) then
+        call obj_mgbf(ig)%anal_to_filt_all2(work_mgbf)
+     else
+        call obj_mgbf(ig)%anal_to_filt_all(work_mgbf)
+     endif
+     deallocate(work_mgbf)
+
+! Apply horizontal smoother for number of horizontal scales
+     call obj_mgbf(ig)%filtering_procedure(obj_mgbf(ig)%mgbf_proc,-1)
+     ii=0
+     do k=1,obj_mgbf(ig)%km_all
+        do j=1-obj_mgbf(ig)%hy,obj_mgbf(ig)%jm+obj_mgbf(ig)%hy
+           do i=1-obj_mgbf(ig)%hx,obj_mgbf(ig)%im+obj_mgbf(ig)%hx
+              ii=ii+1
+              z(ii)=obj_mgbf(ig)%VALL(k,i,j)
+           enddo
+        enddo
+     enddo
 
+! Recursive/Spectral filter-based localization
+  else
+
```

Comment:
for previous recursive branch, could we do the minimal changes and don't touch them if possible, for example, like the call is not added to 4111 in the loop with openmp parallelization.     Any changes that look trivial /ok may cause errors ,especially for call of sub in the opm parallelized loop is very tricky and need much efforts to verify.  So, I suggest to use the original codes for rf branch .

Reply:
Corrected.